pkg/checks/external/ssl_util/main.go

```diff
@@ -17,37 +17,80 @@ import (
 
 var TimeoutSeconds = 10
 
-// CertHandshake performs a basic TLS/SSL handshake on the specified host/port, returning any errors
-func CertHandshake(host, port string) error {
-	log.Info("Testing SSL handshake on host ", host, " over port ", port)
+// CertificatePuller makes an insecure connection to a host and specified port, and returns the SSL certificate if the DNSNames field matches the hostname
+func CertificatePuller(host, port string) (*x509.Certificate, error) {
+	var hostCert *x509.Certificate
+	d := &net.Dialer{
+		Timeout: time.Duration(TimeoutSeconds) * time.Second,
+	}
+
+	conn, err := tls.DialWithDialer(d, "tcp", host+":"+port, &tls.Config{
+		InsecureSkipVerify: true,
+		MinVersion:         tls.VersionTLS12,
+	})
+
+	if err != nil {
+		log.Warnln([]*x509.Certificate{&x509.Certificate{}}, "", err)
+		return hostCert, err
+	}
+
+	defer conn.Close()
+
+	cert := conn.ConnectionState().PeerCertificates
+
+	for _, clientCert := range cert {
+		for _, certDNS := range clientCert.DNSNames {
+			if certDNS == host {
+				hostCert = clientCert
+			}
+		}
+	}
+	return hostCert, err
+}
+
+// HandshakeFromHost performs a TLS handshake using a self signed cert, pulled from the host by the CertificatePuller function, returning any errors
+func HandshakeFromHost(host, port string, cert *x509.Certificate) error {
 	d := &net.Dialer{
 		Timeout: time.Duration(TimeoutSeconds) * time.Second,
 	}
 
+	// read the system cert pool, proceed with a new, empty pool if no certs found
+	rootCAs, _ := x509.SystemCertPool()
+	if rootCAs == nil {
+		rootCAs = x509.NewCertPool()
+	}
+
+	if cert != nil {
+		rootCAs.AddCert(cert)
+	}
+
 	conn, err := tls.DialWithDialer(d, "tcp", host+":"+port, &tls.Config{
 		InsecureSkipVerify: false,
 		MinVersion:         tls.VersionTLS12,
+		RootCAs:            rootCAs,
 	})
 
 	if err != nil {
 		log.Warnln([]*x509.Certificate{&x509.Certificate{}}, "", err)
 		return err
 	}
+
 	defer conn.Close()
 
 	err = conn.Handshake()
 	if err != nil {
 		log.Warn("Unable to complete SSL handshake with host ", host+": ", err)
 		return err
 	}
+
 	log.Info("SSL handshake to host ", host, " completed successfully")
 	return err
 }
```

Comment:
seems like this is duplicated here:
https://github.com/Comcast/kuberhealthy/pull/658/files#diff-86d238c1d056bcbd1fb47e415fdb9ea8R115-R133

wondering if we can pull this out into a seperate function that makes the ssl handshake. Both the `HandshakeFromFile` and `HandshakeFromHost` pulls a cert that gets added to the `RootCAs`

Reply:
seems like that function is right here in `CertHandshake`
https://github.com/Comcast/kuberhealthy/pull/658/files#diff-86d238c1d056bcbd1fb47e415fdb9ea8R137-R162

the only thing missing / different is the `RootCAs:            rootCAs,`

Reply:
I believe this has been updated satisfactorily in the last PR. Resolving.